syntax = "proto3";

package post;

service PostService {
  rpc CreatePost (CreatePostRequest) returns (PostResponse);
  rpc GetPostById (GetPostRequest) returns (PostResponse);
  rpc UpdatePost (UpdatePostRequest) returns (PostResponse);
  rpc DeletePost (DeletePostRequest) returns (EmptyResponse);
  rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);

  rpc ViewPost (ViewPostRequest) returns (EmptyResponse);
  rpc LikePost (LikePostRequest) returns (EmptyResponse);
  rpc AddComment (AddCommentRequest) returns (CommentResponse);
  rpc ListComments (ListCommentsRequest) returns (ListCommentsResponse);
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  int32 creator_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message UpdatePostRequest {
  int32 id = 1;
  string title = 2;
  string description = 3;
  bool is_private = 4;
  repeated string tags = 5;
  int32 requestor_id = 6;
}

message GetPostRequest {
  int32 id = 1;
}

message DeletePostRequest {
  int32 id = 1;
  int32 requestor_id = 2;
}

message ListPostsRequest {
  int32 page = 1;
  int32 size = 2;
  optional int32 creator_id = 3;
}

message PostResponse {
  int32 id = 1;
  string title = 2;
  string description = 3;
  int32 creator_id = 4;
  string created_at = 5;
  string updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
  int32 likes_count = 9;
  int32 views_count = 10;
}

message ListPostsResponse {
  repeated PostResponse posts = 1;
  int32 total = 2;
}

message EmptyResponse {}

message ViewPostRequest {
    int32 post_id = 1;
    int32 user_id = 2;
}

message LikePostRequest {
    int32 post_id = 1;
    int32 user_id = 2;
}

message AddCommentRequest {
    int32 post_id = 1;
    int32 user_id = 2;
    string content = 3;
}

message CommentResponse {
    int32 id = 1;
    int32 post_id = 2;
    int32 user_id = 3;
    string content = 4;
    string created_at = 5;
}

message ListCommentsRequest {
    int32 post_id = 1;
    int32 page = 2;
    int32 size = 3;
}

message ListCommentsResponse {
    repeated CommentResponse comments = 1;
    int32 total = 2;
}